### PR TITLE
feat(planner): support udf

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

as the tile

## Changelog

- New Feature

## Related Issues

Fixes https://github.com/datafuselabs/databend/issues/5647

### CODE REVIEW

File: common/ast/src/parser/mod.rs [hunk 1]
```diff
@@ -52,3 +53,20 @@ pub fn parse_sql<'a>(
         Err(nom::Err::Incomplete(_)) => unreachable!(),
     }
 }
+
+/// Parse udf function into Expr
+pub fn parse_udf<'a>(
```
Comment (leiysky) [inline common/ast/src/parser/mod.rs:58]: I think `parse_expr` is better.

File: query/src/sql/planner/semantic/type_check.rs [hunk 2]
```diff
@@ -446,9 +451,12 @@ impl<'a> TypeChecker<'a> {
     pub async fn resolve_function(
         &mut self,
         func_name: &str,
-        arguments: &[&Expr<'a>],
+        arguments: &[&Expr<'_>],
         _required_type: Option<DataTypeImpl>,
     ) -> Result<(Scalar, DataTypeImpl)> {
+        if !is_builtin_function(func_name) {
+            return self.resolve_udf(func_name, arguments).await;
+        }
```
Comment (leiysky) [inline query/src/sql/planner/semantic/type_check.rs:459]: We should move this check to https://github.com/datafuselabs/databend/blob/83884cd152d43c5d77907262a4a58f2d8e0be083/query/src/sql/planner/semantic/type_check.rs#L259-L268

File: common/ast/src/parser/mod.rs (hunk 1 above)
Comment (xudong963) [reply common/ast/src/parser/mod.rs]: done

File: query/src/sql/planner/semantic/type_check.rs (hunk 2 above)
Comment (xudong963) [reply query/src/sql/planner/semantic/type_check.rs]: done

File: query/src/sql/planner/semantic/type_check.rs [hunk 3]
```diff
@@ -933,4 +940,284 @@ impl<'a> TypeChecker<'a> {
...
+            })?;
+            self.resolve(&udf_expr, None).await
+        } else {
+            Err(ErrorCode::SemanticError(
+                "No function matches the given name.",
+            ))
+        };
+    }
+
+    fn clone_expr_with_replacement<F>(
```
Comment (leiysky) [inline query/src/sql/planner/semantic/type_check.rs:988]: It's better to do this with a unified visitor, but we can open another issue for this.

**Review by leiysky — APPROVED** [review]

File: query/src/sql/planner/semantic/type_check.rs (hunk 3 above)
Comment (xudong963) [reply query/src/sql/planner/semantic/type_check.rs:988]: Ok

**Pull request merged**